### PR TITLE
Add Dependency-Injection Higher-Order Component

### PR DESCRIPTION
**Pull request by shannonmoeller**
RFC: https://github.com/fusionjs/rfcs/pull/11
Depends on: https://github.com/fusionjs/fusion-core/pull/290

### CODE REVIEW

File: src/injector.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+export function registerInjector(app: FusionApp) {
+  function inject(deps: Dependencies): Services {
+    return getServices(app, deps);
+  }
+
+  const injectorPlugin = createPlugin({
+    middleware: () => (ctx, next) => {
+      ctx.element = (
```
Comment (ganemone) [inline src/injector.js]: This should check for `ctx.element` before wrapping it. `ctx.element` is only present on server side renders

**Comment by shannonmoeller** [comment]
@ganemone I've added support for the legacy api if the new context api isn't available. This is easily removed once support for older versions of react is dropped.

**Comment by shannonmoeller** [comment]
That coverage report is wrong.

File: src/injector.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import * as React from 'react';
+import FusionApp, {createPlugin} from 'fusion-core';
+import PropTypes from 'prop-types';
+
+type Dependencies = {[string]: any};
```
Comment (ganemone) [inline src/injector.js]: I think we can refine these types a bit, but this is ok for now

**Comment by ganemone** [comment]
!merge

**Comment by rasmusprentow** [comment]
Nice! This actually solves this issue https://github.com/fusionjs/fusion-react/issues/106.

Good Job!

**Pull request merged**